tiledb/sm/serialization/query_plan.cc

```diff
@@ -0,0 +1,314 @@
+/**
+ * @file query_plan.cc
+ *
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2023 TileDB, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file implements serialization for a query plan
+ */
+
+// clang-format off
+#ifdef TILEDB_SERIALIZATION
+#include <capnp/compat/json.h>
+#include <capnp/message.h>
+#include <capnp/serialize.h>
+#include "tiledb/sm/serialization/capnp_utils.h"
+#endif
+// clang-format on
+
+#include "tiledb/sm/query_plan/query_plan.h"
+#include "tiledb/common/thread_pool.h"
+#include "tiledb/sm/config/config.h"
+#include "tiledb/sm/enums/serialization_type.h"
+#include "tiledb/sm/serialization/query.h"
+#include "tiledb/sm/serialization/query_plan.h"
+
+using namespace tiledb::common;
+
+namespace tiledb::sm::serialization {
+
+#ifdef TILEDB_SERIALIZATION
+
+void query_plan_request_to_capnp(
+    capnp::QueryPlanRequest::Builder& builder,
+    const Config& config,
+    Query& query) {
+  auto config_builder = builder.initConfig();
+  throw_if_not_ok(config_to_capnp(config, &config_builder));
+
+  auto query_builder = builder.initQuery();
+  throw_if_not_ok(query_to_capnp(query, &query_builder, true));
+}
+
+void query_plan_request_from_capnp(
+    const capnp::QueryPlanRequest::Reader& reader,
+    ThreadPool& compute_tp,
+    Query& query) {
+  if (reader.hasQuery()) {
```

Comment:
We should probably deserialize the config as well?

Reply:
@ypatia Exactly right. The config object in all of the request messages is for the benefit of the cloud rest server.